lib/spack/spack/cmd/buildcache.py

```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
         spack.binary_distribution._url_generate_package_index(url, tmpdir)
 
     if update_keys:
-        try:
-            with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpdir:
-                spack.binary_distribution.generate_key_index(url, tmpdir)
-        except spack.binary_distribution.CannotListKeys as e:
-            # Do not error out if listing keys went wrong. This usually means that the _gpg path
-            # does not exist. TODO: distinguish between this and other errors.
-            tty.warn(f"did not update the key index: {e}")
+        mirror_update_keys(mirror)
+
+
+def mirror_update_keys(mirror: spack.mirrors.mirror.Mirror):
+    url = mirror.push_url
+    try:
+        with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpdir:
+            spack.binary_distribution.generate_key_index(url, tmpdir)
+    except spack.binary_distribution.CannotListKeys as e:
+        # Do not error out if listing keys went wrong. This usually means that the _gpg path
+        # does not exist. TODO: distinguish between this and other errors.
+        tty.warn(f"did not update the key index: {e}")
+
+
+def read_concrete_hashes(source: str) -> List[str]:
+    """Read all of the concrete hashes from a given source"""
+    try:
+        env = ev.environment_from_name_or_dir(source)
+        return list(env.all_hashes())
+    except ev.SpackEnvironmentError as e:
+        raise spack.error.SpackError(f"Failed to read specs from source: {source}") from e
+
+
+def update_view(
+    mirror: spack.mirrors.mirror.Mirror,
+    update_mode: ViewUpdateMode,
+    *sources: str,
+    name: Optional[str] = None,
+    update_keys: bool = False,
+):
+    """update a buildcache view index"""
+    # OCI images do not support views.
+    try:
+        spack.oci.oci.image_from_mirror(mirror)
+        raise spack.error.SpackError("OCI build caches do not support index views")
+    except ValueError:
+        pass
+
+    if update_mode == ViewUpdateMode.APPEND:
+        tty.warn(
+            "Appending to a view index does not guarantee idempotent write when contending "
+            "with multiple writers. This feature is meant to be used by a single process."
+        )
+
+    # Otherwise, assume a normal mirror.
+    url = mirror.push_url
+
+    if (name and mirror.push_view) and not name == mirror.push_view:
+        tty.warn(
+            (
+                f"Updating index view with name ({name}), which is different than "
+                f"the configured name ({mirror.push_view}) for the mirror {mirror.name}"
+            )
+        )
+
+    name = name or mirror.push_view
+    assert name
+
+    url_and_version = spack.binary_distribution.MirrorMetadata(
+        url, spack.binary_distribution.CURRENT_BUILD_CACHE_LAYOUT_VERSION, name
+    )
```

Comment:
```suggestion
    metadata = spack.binary_distribution.MirrorMetadata(
        url, spack.binary_distribution.CURRENT_BUILD_CACHE_LAYOUT_VERSION, name
    )
```
the variable name is misleading since we've refactored things.